pkg/sql/exec/error.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2019 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package exec
+
+import (
+	"bufio"
+	"fmt"
+	"runtime/debug"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/exec/coldata"
+	"github.com/pkg/errors"
+)
+
+const (
+	panicLineSubstring = "runtime/panic.go"
+	execPackagePrefix  = "github.com/cockroachdb/cockroach/pkg/sql/exec"
+)
+
+// CatchVectorizedRuntimeError executes operation, catches a runtime error if
+// it is coming from exec package, and returns it. If an error occurs that is
+// not from exec package, it is not recovered from.
+func CatchVectorizedRuntimeError(operation func()) (retErr error) {
+	defer func() {
+		if err := recover(); err != nil {
+			stackTrace := string(debug.Stack())
+			scanner := bufio.NewScanner(strings.NewReader(stackTrace))
+			panicLineFound := false
+			for scanner.Scan() {
+				if strings.Contains(scanner.Text(), panicLineSubstring) {
+					panicLineFound = true
+					break
+				}
+			}
+			if !panicLineFound {
+				panic(fmt.Sprintf("panic line %q not found in the stack trace\n%s", panicLineSubstring, stackTrace))
+			}
+			if scanner.Scan() {
+				if strings.HasPrefix(strings.TrimSpace(scanner.Text()), execPackagePrefix) {
+					// We only want to catch runtime errors coming from the exec package.
+					if e, ok := err.(error); ok {
+						retErr = e
+					} else {
+						// panic occurred with an object that is not error.
+						retErr = errors.New(fmt.Sprintf("%v", err))
+					}
+				} else {
+					// Do not recover from the panic not related to the vectorized
+					// engine.
+					panic(err)
+				}
+			} else {
+				panic(fmt.Sprintf("unexpectedly there is no line below the panic line in the stack trace\n%s", stackTrace))
+			}
+		} else {
+			// No panic happened, so the operation must have been executed
+			// successfully.
+			retErr = nil
```

Comment:
This line is unnecessary, because the zero value for `error` is `nil` - which means `retErr` starts off as `nil` anyway.

Reply:
Ok, removed. I thought of making this assignment as being more explicit which might make reading the code easier. (I still kept the comment though.)